File: src/ru/hudspeed/pro/CarRadarBaseUpdateNotifier.java
package ru.gpsantiradar.app;

final class CarRadarBaseUpdateNotifier {
    interface Source {
        void addListener(RadarBaseUpdater.Listener listener, boolean replayLatest);
        void removeListener(RadarBaseUpdater.Listener listener);
    }

    interface MessageSink {
        void show(RadarBaseUpdateState state);
    }

    private final Source source;
    private final MessageSink messageSink;
    private boolean started;

    private final RadarBaseUpdater.Listener listener =
            new RadarBaseUpdater.Listener() {
                @Override public void onRadarBaseUpdate(RadarBaseUpdateState state) {
                    if (started && state != null
                            && state.status != RadarBaseUpdateState.Status.IDLE) {
                        messageSink.show(state);
                    }
                }
            };

    CarRadarBaseUpdateNotifier(RadarBaseUpdater updater, MessageSink messageSink) {
        this(new SharedSource(updater), messageSink);
    }

    CarRadarBaseUpdateNotifier(Source source, MessageSink messageSink) {
        if (source == null) throw new IllegalArgumentException("source is required");
        if (messageSink == null) {
            throw new IllegalArgumentException("messageSink is required");
        }
        this.source = source;
        this.messageSink = messageSink;
    }

    void start() {
        if (started) return;
        started = true;
        source.addListener(listener, true);
    }

    void stop() {
        if (!started) return;
        started = false;
        source.removeListener(listener);
    }

    private static final class SharedSource implements Source {
        private final RadarBaseUpdater updater;

        SharedSource(RadarBaseUpdater updater) {
            if (updater == null) {
                throw new IllegalArgumentException("updater is required");
            }
            this.updater = updater;
        }

        @Override public void addListener(
                RadarBaseUpdater.Listener listener, boolean replayLatest) {
            updater.addListener(listener, replayLatest);
        }

        @Override public void removeListener(RadarBaseUpdater.Listener listener) {
            updater.removeListener(listener);
        }
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/CarRadarBaseUpdateNotifier.kt
package ru.gpsantiradar.app

class CarRadarBaseUpdateNotifier {
    interface Source {
        fun latestState(): RadarBaseUpdateState?
        fun addListener(listener: RadarBaseUpdater.Listener, replayLatest: Boolean)
        fun removeListener(listener: RadarBaseUpdater.Listener)
    }
    fun interface MessageSink { fun show(state: RadarBaseUpdateState) }
    private val source: Source
    private val messageSink: MessageSink
    private var started = false
    private var terminalSequenceAtStart = Long.MIN_VALUE
    private val listener = object : RadarBaseUpdater.Listener {
        override fun onRadarBaseUpdate(state: RadarBaseUpdateState) {
            if (started && state.status != RadarBaseUpdateState.Status.IDLE) {
                if (state.isTerminal() && state.sequence <= terminalSequenceAtStart) return
                messageSink.show(state)
            }
        }
    }
    constructor(updater: RadarBaseUpdater?, messageSink: MessageSink?) : this(SharedSource(requireNotNull(updater) { "updater is required" }), messageSink)
    constructor(source: Source?, messageSink: MessageSink?) {
        this.source = requireNotNull(source) { "source is required" }
        this.messageSink = requireNotNull(messageSink) { "messageSink is required" }
    }
    fun start() {
        if (started) return
        val latest = source.latestState()
        terminalSequenceAtStart = if (latest?.isTerminal() == true) latest.sequence else Long.MIN_VALUE
        started = true
        source.addListener(listener, true)
    }
    fun stop() {
        if (!started) return
        started = false
        source.removeListener(listener)
    }
    private class SharedSource(private val updater: RadarBaseUpdater) : Source {
        override fun latestState() = updater.latestState()
        override fun addListener(listener: RadarBaseUpdater.Listener, replayLatest: Boolean) = updater.addListener(listener, replayLatest)
        override fun removeListener(listener: RadarBaseUpdater.Listener) = updater.removeListener(listener)
    }
}
